core/vm/contracts.go

```
@@ -1203,3 +1216,313 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
+			the last 4 is the maxNumber of topics
+		0000000000000000000000000000000000000000000000000000000000000060
+		00000000000000000000000000000000000000000000000000000000000000c0
+		0000000000000000000000000000000000000000000000000000000000000002
+		dce721dc2d078c030530aeb5511eb76663a705797c2a4a4d41a70dddfb8efca9
+		000000000000000000000000000000000000000000000000000000000000000
+
+		if maxDataLen % 32 != 0
+			data_len= 32 + (32 - maxDataLen % 32) + maxDatalen
+		if maxDataLen % 32 == 0
+			data_len= 32 + maxDatalen
+		Note:
+			the first 32 is the length of data which generated after ABIPack(address,topics,data)
+		00000000000000000000000000000000000000000000000000000000000000c0
+		0000000000000000000000000000000000000000000000000000000000000002
+		0000000000000000000000000000000000000000000000000000000000000001
+		0000000000000000000000000000000000000000000000000000000000000060
+		0000000000000000000000000000000000000000000000000000000000000028
+		bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
+		bbbbbbbbbbbbbbbb000000000000000000000000000000000000000000000000
+	*/
+
+	return gasUsed
+}
+
+func (c *crossChainCall) Run(input []byte) ([]byte, error) {
+	return nil, nil
+}
+
+func (c *crossChainCall) RunWith(env *PrecompiledContractCallEnv, input []byte) ([]byte, uint64, error) {
+	if len(input) != CrossChainCallInputLength {
+		return nil, 0, ErrInvalidCrossChainCallInputLength
+	}
+
+	if env.evm.ChainConfig().ExternalCall.EnableBlockNumber == nil {
+		return nil, 0, ErrDisableExternalCall
+	}
+
+	if env.evm.Context.BlockNumber.Cmp(env.evm.ChainConfig().ExternalCall.EnableBlockNumber) == -1 {
+		return nil, 0, ErrExternalCallNoActive
+	}
+
+	ctx := context.Background()
+	if bytes.Equal(input[0:4], getLogByTxHashId) {
+
+		client := env.evm.ExternalCallClient()
+		var list *CrossChainCallResults
+
+		if client == nil {
```

when `ExternalCallClient` is nil, I think it's better to just return an explicit error? For a call to `crossChainCall` to finally succeed, at least quorum of validators should have set `ExternalCallClient`. In this regard I think `VerifyCrossChainCall` is not needed.

---

We need to allow ordinary nodes to still have the ability to execute external-call-tx, so as to ensure that the entire state-sync process will not be affected.

How ordinary nodes handle this kind of transaction？

- Node receives the block through p2p network when state-sync.
- Node fetchs all the external-call result from block.uncles.
- Node re-executes all transactions of the block.
  - The transaction set the external-call result at evm using `vmenv.SetCrossChainCallResultsByUncles(block.Uncles(), tx)` to provides the ability to execute external-call transactions at `crossChainCall` precompile contract for normal node.
  - The crossChainCall fetchs the external-call result from `env.evm.Interpreter().CrossChainCallResults()[Idx]` when the client is nil to continue to execute the transaction.

---

Add new flag IsStateSync to deal with this case: https://github.com/QuarkChain/go-ethereum/pull/101/commits/52dde717fd6638bf3586f04f56992046beed1fe3

---

Is `SetCrossChainCallResultsByUncles` necessary? What about just passing the additional Uncles etc and do calculations inside the precompile ?

---

The function named `SetCrossChainCallResultsByUncles()` has been removed.